### PR TITLE
add tests for install & uninstall

### PR DESCRIPTION
**Pull request by TitanSnow**
add tests for install & uninstall, with/without `sudo`

I also fix the `$(xmake)` so that it will be always correct. See details at commit message

Have some problem on mac os on travis-ci

### CODE REVIEW

**Comment by waruqi** [comment]
@TitanSnow Using `$(xmake)` has some of the following issues:

* If the path has spaces (on windows. .e.g `"C:\Program Files\xmake\xmake.exe"`), double quotes are required, for example:

```
os.run("\"$(xmake)\" f -c")
```

* If users don't see the document, they don't know if they have this variable: `$(xmake)`, then they will still run it directly. `os.run("xmake")`
 
So, I'm going to remove `$(xmake)` and use `os.run("xmake")` directly and add the executable path of xmake to `PATH` env. for example:

```lua
function main()
        ..
        os.addenv("PATH", xmake._EXECUTABLE_DIR)
```

I think it's easier to use, and users don't have to care about the path of xmake.

**Comment by TitanSnow** [comment]
Great! :+1:

**Comment by TitanSnow** [comment]
There is a situation that needs `$(xmake)`. That is, `sudo $(xmake)`. `sudo` will delete environment var so that the path of xmake must be known

**Comment by TitanSnow** [comment]
`luacov.runner` not found on macos. Wired

**Comment by waruqi** [comment]
@TitanSnow 
> There is a situation that needs $(xmake). That is, sudo $(xmake). sudo will delete environment var so that the path of xmake must be known.

You can use `sudo -E xmake` or `os.sudo(os.run, "xmake f -c")`(in [os.lua](https://github.com/tboox/xmake/blob/master/xmake/core/sandbox/modules/os.lua)).

Pass `-E` argument will preserve user environment when running `sudo` command. So I think `$(xmake)` is not necessary.

**Comment by TitanSnow** [comment]
no use

```console
$ xmake l
> os.getenv("PATH")
/home/ts/.local/share/xmake:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/ts/.local/bin
> os.sudo(os.exec,"sh -c 'echo $PATH'")
[sudo] password for ts: 
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
> os.exec("sudo -E sh -c 'echo $PATH'")
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
```

**Comment by waruqi** [comment]
@TitanSnow 

It works on my mac machine. 

```console
xmake l
> os.addenv("PATH", "/tmp/test")
> os.getenv("PATH")
/tmp/test:/Users/ruki/.cargo/bin:/Users/ruki/.rvm/gems/ruby-2.2.2/bin:/Users/ruki/.rvm/gems/ruby-2.2.2@global/bin:/Users/ruki/.rvm/rubies/ruby-2.2.2/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/ruki/.rvm/bin
> os.exec("sudo -E xmake l --root -c 'print(os.getenv(\"PATH\"))'")
/tmp/test:/Users/ruki/.cargo/bin:/Users/ruki/.rvm/gems/ruby-2.2.2/bin:/Users/ruki/.rvm/gems/ruby-2.2.2@global/bin:/Users/ruki/.rvm/rubies/ruby-2.2.2/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/ruki/.rvm/bin
> os.sudo(os.exec, "xmake l --root -c 'print(os.getenv(\"PATH\"))'")
/tmp/test:/Users/ruki/.cargo/bin:/Users/ruki/.rvm/gems/ruby-2.2.2/bin:/Users/ruki/.rvm/gems/ruby-2.2.2@global/bin:/Users/ruki/.rvm/rubies/ruby-2.2.2/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/ruki/.rvm/bin
> os.exec("sudo -E sh -c 'echo $PATH'")
/tmp/test:/Users/ruki/.cargo/bin:/Users/ruki/.rvm/gems/ruby-2.2.2/bin:/Users/ruki/.rvm/gems/ruby-2.2.2@global/bin:/Users/ruki/.rvm/rubies/ruby-2.2.2/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/ruki/.rvm/bin
> os.sudo(os.exec,"sh -c 'echo $PATH'")
/tmp/test:/Users/ruki/.cargo/bin:/Users/ruki/.rvm/gems/ruby-2.2.2/bin:/Users/ruki/.rvm/gems/ruby-2.2.2@global/bin:/Users/ruki/.rvm/rubies/ruby-2.2.2/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/ruki/.rvm/bin
```

**Comment by TitanSnow** [comment]
I found some info on [stackoverflow](https://stackoverflow.com/questions/257616). Seems ubuntu's behavior on `sudo` is different from others

**Comment by waruqi** [comment]
@TitanSnow I am going to try add `detect.tool.find_sudo` and `privilege.sudo` modules to handle these differences

**Comment by waruqi** [comment]
@TitanSnow I have removed the `$(xmake)` variable and and even if `sudo` cannot inherit the parent environment, the path of `xmake` is always in the `$PATH` because I'll add it's directory to `$PATH` in the main entry function.

```lua
function main.init()

    -- add the directory of the program file (xmake) to $PATH environment
    local programfile = os.programfile()
    if programfile and os.isfile(programfile) then
        os.addenv("PATH", path.directory(programfile))
    else
        os.addenv("PATH", os.programdir())
    end
```

**Comment by TitanSnow** [comment]
So how to deal with build error on travis-ci? The trouble is **`sudo` cannot find xmake** not xmake cannot find itself

**Comment by waruqi** [comment]
@TitanSnow Can you give me your failed travis-ci links?

**Comment by TitanSnow** [comment]
Below :arrow_down: :smiley:

**Comment by waruqi** [comment]
@TitanSnow 

> So how to deal with build error on travis-ci? 

This seems like another problem about `luacov.runner`. I will look at it.

> The trouble is sudo cannot find xmake not xmake cannot find itself

I think xmake can find itself from `os.programfile()` and I see [travis](https://travis-ci.org/tboox/xmake/builds/235516658) has passed

**Comment by TitanSnow** [comment]
> I think xmake can find itself from os.programfile() and I see travis has passed

That is because 962b99c :sweat:

**Comment by waruqi** [comment]
@TitanSnow Ok, I will look at it.

**Comment by waruqi** [comment]
@TitanSnow I'm doing some tests on [travis-ci](https://travis-ci.org/tboox/xmake/jobs/235602647)

The path of xmake already exists when running `sudo`

But xmake still not found,I don't know why. 😢 

```console
$ which xmake
/home/travis/bin/xmake

$ sudo sh -c "echo $PATH"
/home/travis/bin:/home/travis/.local/bin:/home/travis/.pyenv/shims:/home/travis/.phpenv/shims:/home/travis/gopath/bin:/home/travis/.gimme/versions/go1.7.linux.amd64/bin:/home/travis/.local/bin:/opt/pyenv/bin:/opt/python/2.7.12/bin:/opt/python/3.5.2/bin:/opt/python/2.6.9/bin:/opt/python/3.2.6/bin:/opt/python/3.3.6/bin:/opt/python/3.4.4/bin:/opt/python/pypy-5.4.1/bin:/opt/python/pypy3-2.4.0/bin:/usr/local/phantomjs/bin:/usr/local/phantomjs:/home/travis/perl5/perlbrew/bin:/home/travis/.nvm/versions/node/v4.1.2/bin:/usr/local/maven-3.1.1/bin:/home/travis/.kiex/elixirs/elixir-1.0.4/bin:/home/travis/.kiex/bin:/usr/local/gradle/bin:/usr/local/clang-3.5.0/bin:/home/travis/.rvm/gems/ruby-2.3.1/bin:/home/travis/.rvm/gems/ruby-2.3.1@global/bin:/home/travis/.rvm/rubies/ruby-2.3.1/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/lib/jvm/java-9-oracle/bin:/usr/lib/jvm/java-9-oracle/db/bin:/home/travis/.phpenv/bin:/home/travis/.rvm/bin:/home/travis/.lua:/home/travis/.local/bin:/home/travis/build/tboox/xmake/install/luarocks/bin

$ sudo sh -c "xmake --root lua versioninfo"
sh: 1: xmake: not found
```

**Comment by TitanSnow** [comment]
@waruqi The problem is funny: double quote will expand the vars inside :smiley_cat: 

```console
$ sudo sh -c 'echo $PATH'   
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
$ sudo sh -c "echo $PATH"
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/ts/.local/bin
```

**Comment by waruqi** [comment]
@TitanSnow Oh, Thanks! 👍  But everything is normal on my vbox machines(ubuntu, mint, lubuntu ..) 

and why would xmake be installed in this directory `/home/travis/bin`? 

I think this should be an installation problem,  xmake should be installed in `/usr/local/bin` by default.

```console
$ which xmake
/home/travis/bin/xmake

$ sh -c 'echo $PATH'
/home/travis/bin:/home/travis/.local/bin:/home/travis/.pyenv/shims:/home/travis/.phpenv/shims:/home/travis/gopath/bin:/home/travis/.gimme/versions/go1.7.linux.amd64/bin:/home/travis/.local/bin:/opt/pyenv/bin:/opt/python/2.7.12/bin:/opt/python/3.5.2/bin:/opt/python/2.6.9/bin:/opt/python/3.2.6/bin:/opt/python/3.3.6/bin:/opt/python/3.4.4/bin:/opt/python/pypy-5.4.1/bin:/opt/python/pypy3-2.4.0/bin:/usr/local/phantomjs/bin:/usr/local/phantomjs:/home/travis/perl5/perlbrew/bin:/home/travis/.nvm/versions/node/v4.1.2/bin:/usr/local/maven-3.1.1/bin:/home/travis/.kiex/elixirs/elixir-1.0.4/bin:/home/travis/.kiex/bin:/usr/local/gradle/bin:/usr/local/clang-3.5.0/bin:/home/travis/.rvm/gems/ruby-2.3.1/bin:/home/travis/.rvm/gems/ruby-2.3.1@global/bin:/home/travis/.rvm/rubies/ruby-2.3.1/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/lib/jvm/java-9-oracle/bin:/usr/lib/jvm/java-9-oracle/db/bin:/home/travis/.phpenv/bin:/home/travis/.rvm/bin:/home/travis/.lua:/home/travis/.local/bin:/home/travis/build/tboox/xmake/install/luarocks/bin

$ sudo sh -c 'echo $PATH'
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
```

**Comment by waruqi** [comment]
And it will be installed to `/Users/ruki/.cargo` when I run `./scripts/get.sh __local__` on my machine.

```console
...
installing to /Users/ruki/.cargo ...      
```

I think it's a problem.

**Comment by TitanSnow** [comment]
@waruqi :sweat: That is my install script did that. Install to there does not need root access

See `scripts/get.sh`

**Comment by waruqi** [comment]
@TitanSnow But this maybe overwrite the files other software and `sudo xmake` not found.

**Comment by TitanSnow** [comment]
I have an idea, that is

1. install to a certain path that not require root
1. append path to `~/.bashrc`
1. give a tip to let user run `source ~/.bashrc` in current session -> path ok
1. after restarting, shell loads rc -> path ok

**Comment by waruqi** [comment]
@TitanSnow So how does it work on travis-ci or `sudo xmake` if need user to run `source ~/.bashrc`?

**Comment by TitanSnow** [comment]
@waruqi The environment var PATH is loaded from `~/.bashrc`. You could open it and have a look, there might be a line like `export PATH=...`

So `source ~/.bashrc` is a **command** to reload the rc. And next time open terminal, there is no need

You could see `.travis.yml` line 24. To install luarocks I used source so there is no problem with no user

**Comment by waruqi** [comment]
@TitanSnow  Can `sudo` find xmake?

**Comment by waruqi** [comment]
I improved `sudo` module, use `sudo PATH=$PATH command` instead of `sudo -E command` to  preserve parent environment. It's more generic.

```lua
unction _sudo(runner, cmd, ...)

    -- find sudo
    local program = find_sudo()
    assert(program, "sudo not found!")

    -- run it with administrator permission and preserve parent environment
    runner(program .. " PATH=" .. os.getenv("PATH") .. " " .. cmd, ...)
end
```

**Comment by TitanSnow** [comment]
@waruqi Did you deal with `PATH` with spaces and quotes? For example `PATH=/home/awd/i am space:/home/awd/i'm quote`?

**Comment by waruqi** [comment]
> Did you deal with PATH with spaces and quotes? For example PATH=/home/awd/i am space:/home/awd/i'm quote?

I have fixed it.

File: core/src/xmake/machine.c [hunk 1]
```diff
@@ -260,75 +291,73 @@ static tb_size_t xm_machine_get_program_file(xm_machine_impl_t* impl, tb_char_t*
...
+
+            // set _EXECUTABLE_PATH
+            lua_pushstring(impl -> lua, data);
+            lua_setglobal(impl -> lua, "_EXECUTABLE_PATH");
+
+            if (!ok)
             {
                 // get the directory
                 while (size-- > 0)
                 {
```
Comment (waruqi) [inline core/src/xmake/machine.c]: I don't think it's necessary to write the following code three times.

```c
                while (size-- > 0)
                {
                    if (data[size] == '/')
                    {
                        data[size] = '\0';
                        break;
                    }
                }
```

File: xmake/core/_xmake_main.lua [hunk 2]
```diff
@@ -31,12 +31,13 @@ xmake._NULDEV           = _NULDEV
...
 xmake._VERSION_SHORT    = _VERSION_SHORT
 xmake._PROGRAM_DIR      = _PROGRAM_DIR
-xmake._PROGRAM_FILE     = _PROGRAM_FILE
+xmake._PROGRAM_FILE     = _EXECUTABLE_PATH
 xmake._PROJECT_DIR      = _PROJECT_DIR
 xmake._CORE_DIR         = _PROGRAM_DIR .. "/core"
 xmake._TOOLS_DIR        = _PROGRAM_DIR .. "/tools"
 xmake._TEMPLATES_DIR    = _PROGRAM_DIR .. "/templates"
 xmake._PROJECT_FILE     = "xmake.lua"
+xmake._EXECUTABLE_PATH  = _EXECUTABLE_PATH
```
Comment (waruqi) [inline xmake/core/_xmake_main.lua]: I think it's enough to use `xmake._PROGRAM_FILE     = _PROGRAM_FILE`, There's no need to add a redundant variable

File: core/src/xmake/machine.c (hunk 1 above)
Comment (waruqi) [inline core/src/xmake/machine.c]: I think you can use this [implementation in dev](https://github.com/tboox/xmake/blob/dev/core/src/xmake/machine.c) to replace it directly. 😸

**Comment by TitanSnow** [comment]
Done

File: xmake/modules/privilege/sudo.lua [hunk 3]
```diff
@@ -37,8 +37,9 @@ function _sudo(runner, cmd, ...)
     local program = find_sudo()
     assert(program, "sudo not found!")
 
+    -- FIXME: deal with quotes in `os.getenv("PATH")`
     -- run it with administrator permission and preserve parent environment
-    runner(program .. " PATH=\"" .. os.getenv("PATH") .. "\" " .. cmd, ...)
+    runner(program .. " env PATH=\"" .. os.getenv("PATH") .. "\" " .. cmd, ...)
```
Comment (waruqi) [inline xmake/modules/privilege/sudo.lua]: The effect that do not add `env` seems to be the same, what is the difference?

Comment (TitanSnow) [inline xmake/modules/privilege/sudo.lua]:
```console
$ ls ~/awd/build
awd
$ sudo PATH=$HOME/awd/build:$PATH awd
sudo: awd: command not found
$ sudo env PATH=$HOME/awd/build:$PATH awd
hello world!
```

And it makes the build pass

File: makefile [hunk 4]
```diff
@@ -61,15 +61,17 @@ install:
 	@# install the xmake directory
 	@cp -r xmake/* $(xmake_dir_install)
 	@# make the xmake loader
+ifneq ($(PLAT),linux)
```
Comment (waruqi) [inline makefile]: What's this? Although we can get the executable path on linux, macOS and windows. But maybe we cannot get it successfully on some linux distributions.  So I think we still need `export XMAKE_PROGRAM_DIR=/xxx`

**Comment by waruqi** [comment]
Great! 😄

**Comment by waruqi** [comment]
@TitanSnow Can you help me to improve `get.sh` to uninstall xmake? Thanks! 😄 

for example:

```console
$ ./scripts/get.sh __uninstall__
```

**Comment by TitanSnow** [comment]
ok :ok_hand: I'm also going to add a arg to specify install prefix

**Comment by TitanSnow** [comment]
See 8ffb52a4d9bf26f240f6a1a8252b3739047ae8d7

I have impled `__uninstall__` and prefix specification
```console
$ scripts/get.sh __uninstall__
$ env prefix=/usr scripts/get.sh
```

**Comment by waruqi** [comment]
@TitanSnow  👍  

```diff
+    # uninstall
+    makefile=$(remote_get_content https://github.com/tboox/xmake/raw/master/makefile)
+    while which xmake >/dev/null 2>&1
```

I think we can check makefile in the current directory first to faster uninstalling.  for example:

```console
$ cd xmake
$ ./scripts/get.sh __uninstall__
```

**Comment by TitanSnow** [comment]
But how could we know the makefile found in local is xmake's not other?

**Comment by waruqi** [comment]
@TitanSnow Ok. 👌  Can you create a new pr for this commit?

**Comment by TitanSnow** [comment]
Matte, I have a plan of remote rolling pre-build on installation on Windows. That is, the install script get pre-build binary from appveyor rolling build so that no need to build locally to get newest

**Comment by waruqi** [comment]
Good idea! 👍  But how can you get pre-build binary file from appveyor?

**Comment by TitanSnow** [comment]
You could have a look on that branch

**Comment by waruqi** [comment]
@TitanSnow Ok. 👍 😄

**Pull request merged**